src/types/path-set.ts

```diff
@@ -1,120 +1,248 @@
-        const hop = new Hop();
-        type & TYPE_ACCOUNT && (hop.account = AccountID.fromParser(parser));
-        type & TYPE_CURRENCY && (hop.currency = Currency.fromParser(parser));
-        type & TYPE_ISSUER && (hop.issuer = AccountID.fromParser(parser));
-        return hop;
-      },
-    },
-    toJSON() {
-      const type = this.type();
-      const ret = <any>{};
-      type & TYPE_ACCOUNT && (ret.account = this.account.toJSON());
-      type & TYPE_ISSUER && (ret.issuer = this.issuer.toJSON());
-      type & TYPE_CURRENCY && (ret.currency = this.currency.toJSON());
-      return ret;
-    },
-    type() {
-      let type = 0;
-      this.issuer && (type += TYPE_ISSUER);
-      this.account && (type += TYPE_ACCOUNT);
-      this.currency && (type += TYPE_CURRENCY);
-      return type;
-    },
-  },
-  undefined
-);
-
-const Path = makeClass(
-  {
-    inherits: Array,
-    statics: {
-      from(value) {
-        return ensureArrayLikeIs(Path, value).withChildren(Hop);
-      },
-    },
-    toJSON() {
-      return this.map((k) => k.toJSON());
-    },
-  },
-  undefined
-);
-
-const PathSet = makeClass(
-  {
-    mixins: SerializedType,
-    inherits: Array,
-    statics: {
-      from(value) {
-        return ensureArrayLikeIs(PathSet, value).withChildren(Path);
-      },
-      fromParser(parser) {
-        const pathSet = new this();
-        let path;
-        while (!parser.end()) {
-          const type = parser.readUInt8();
-          if (type === PATHSET_END_BYTE) {
-            break;
-          }
-          if (type === PATH_SEPARATOR_BYTE) {
-            path = null;
-            continue;
-          }
-          if (!path) {
-            path = new Path();
-            pathSet.push(path);
-          }
-          path.push(Hop.parse(parser, type));
-        }
-        return pathSet;
-      },
-    },
-    toJSON() {
-      return this.map((k) => k.toJSON());
-    },
-    toBytesSink(sink) {
-      let n = 0;
-      this.forEach((path) => {
-        if (n++ !== 0) {
-          sink.put([PATH_SEPARATOR_BYTE]);
-        }
-        path.forEach((hop) => {
-          sink.put([hop.type()]);
-          hop.account && hop.account.toBytesSink(sink);
-          hop.currency && hop.currency.toBytesSink(sink);
-          hop.issuer && hop.issuer.toBytesSink(sink);
-        });
-      });
-      sink.put([PATHSET_END_BYTE]);
-    },
-  },
-  undefined
-);
+/**
+ * The object representation of a Hop, an issuer AccountID, an account AccountID, and a Currency
+ */
+interface HopObject {
+  issuer?: string;
+  account?: string;
+  currency?: string;
+}
+
+/**
+ * Serialize and Deserialize a Hop
+ */
+class Hop extends SerializedTypeClass {
+  constructor(bytes: Buffer) {
+    super(bytes);
+  }
+
+  /**
+   * Create a Hop from a HopObject
+   * 
+   * @param value Either a hop or HopObject to create a hop with
+   * @returns a Hop
+   */
+  static from(value: Hop | HopObject): Hop {
+    if (value instanceof this) {
```

Comment:
I'd prefer `Hop` to be explicit about the type you're checking for and I think it makes your code more readable too. 

The advantage to `this` (AFAICT) is that if you renamed the class you wouldn't need to change this line. I'm not sure that's an overwhelming reason to compromise readability. 

Let me know if you feel strongly.